### New week new protein stuff!

### Hey nathalie! whatsup

Protein structure visualization, comparison and classification


protein functions are linked to structure

antimicrobials (peptide) covering prosthetics for example

we know about a lot of proteins

but we dont know a lot of those mproteins structure

proteins which have a very similar sequence identinity (100%) , their RMSD (structure similairty) is fairly high

However, going to 60% similairty, their structure is quite similar still

**Structure is more conserved than sequence.**

Homologus proteins have same struct

### Format of structure files

how is information in a 3d protein structure given in a PDB file?

one protein can have more than one structure, but each protein has a CODE for that specific protein structure

20 aa, 

regular enzyme is between 200-300 aa

on average one AA has about 10 atoms

enzyme with 300 aa -> 200 * 10 / 300 * 10 = 2000-3000 atoms

you gotta visualize proteins with tools cuz drawing them is bare stress

its gotta be:

- interactive
- draw connections betweens atoms -> pdb file does not say this!
- software gotta be smart and figure stuff out itself

we help the software by telling it how one aa looks like
and tell that one aa is actually connected to another aa!

thats why in PDB file, ATOM NAME is super important so that you know which atom does which thing
so u can draw connections!

### Example

GNZT 

protease which breaks does Hep C virus

pymol can visualize structure

One can draw structure many ways

Wires -> covalent bonds

Sticks -> covalent

wires -> only backbone -> alpha helix and beta strands makes up backbone

cartoon -> backbone + balls for drugs

surface -> outside of protein -> balls for ligand!

### molecular visualizers:

PyMol
Chimera
VMD
JMol
RasMol

pymol is pretty cool

how to compare structures?

visualzation is cool

but we need a metric, a measure of similairty

you can align structs and that is cooo

but you need methods for similairty without alignment of structures


RMSD -> root mean sqaure deviation

place the molecules in space, and measure distances of molecules 

measure distance between x y and z

REUQIRES MOLECULES TO BE ALIGNED!

SUPERIMPOSE THEM ontop of eachother

RMSD can be calculated for a protein, but also for subgroups of proteins, 

and calculate RMSD after admintering drugs on proteins

bigger proteins -> higher RMSD because higher chance of differences. 

Superimositon is an alogrythm which does it while minizmiing RMSD

chalanges is when different protein seq, alignment is supa tricky

this is called INTERMOLECULAR measures (between molecules)



### Hmm i want to just avoid alignment alltogether


create bonds between all atoms in molecule

create a matrix
4x4 if 4 atoms

use the distance matrix to score if the bonds are good or not, and then try again

This is called INTRAMOLECULAR measures (between matrixes)

One of oldest methods: Dali

uses distances between all pairs of C-alpha atoms in proteins

basically distance between amino acids

dali also has pre computed data for a ton of matrixes

consuqence of those measures -> siutations where distances are the same but maybe proteins are not even connected, they get simialr scores

gets similarities in structure but not maybe connections between you know

### 3 -> Fold databases

SCOP

divide in CLASS, a or b or a/b etc etc

CATH

looksat secondary structure a lot


------

|   | G | G | G | A | A | A | U | C | C |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
| G | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 2 | 3 |
| G |   | 0 | 0 | 0 | 0 | 0 | 1 | 2 | 3 |
| G |   |   | 0 | 0 | 0 | 0 | 1 | 2 | 2 |
| A |   |   |   | 0 | 0 | 0 | 1 | 1 | 1 |
| A |   |   |   |   | 0 | 0 | 1 | 1 | 1 |
| A |   |   |   |   |   | 0 | 1 | 1 | 1 |
| U |   |   |   |   |   |   | 0 | 0 | 0 |
| C |   |   |   |   |   |   |   | 0 | 0 |
| C |   |   |   |   |   |   |   |   | 0 |

