# Lecture 5

NGS/RNA-seq

nex-generation seq


## Origins of dna seq

### why
- medicine ->  genetic variation/rare disease/viruses
- biology genotype to phenotype
- crops 

### sanger
- sanger guy was first dude to do dna seq
- used chain termingating inhibs (sanger seq)

### how do we do sang sequencing
- uses automated seq machines
1. dna is fragmented
2. dna polymrase amplified
3. dNTP's are added and build upon fragments, (for each 4 bases)
4. ddNTP's is also added, and terminates the fragments
5. using gel you can read the bases -> light will correspond to sequence

## NGS is new and cool, millions of sequences

### illumina method

- several lanes
- bind to flow cell
- clonal amplification
- creates bridges
- dna polymerase amplifies and creates dna fragment clusters
- ddntps with flourofour attatched -> light emmison is read from microscope 
- then we gotta assemble this stuff

accurate, fast, robust
*but...* cant read more than 150bp

### pacbio sequence
- single molecule real time sequencing (SMRT)

up to 70kb
no PCR bias

high error rate (compared to the rest)
high cost
no amplifications so need a lot of input material

### nanopore seq

- dna helix is unwound
- dna is fed thru a pore
- there is electric charge in memrbane
- when the dna goes thru, electric charge changes
- change measured and relates to a specific sequence

### ngs important stuff

read length
errors
library type
cost
turnaround time
multiplexing?

### read length
illumina 150kb
pacbio minION 50kb

short are good for re-sequencing
longer reads, LTR, chromosomal reargments

### errors
- there is a lot of error percents
- nanopore is error prone
- illumina is pretty good
- error correcting is a thing

### multiplexing

using an illumina flowcell, can create a barcode

### refrence genome
what would we expect from a genome, can map from it


## RNA

### Transciprtome
- represents all the RNA expressed
- genomes are stable, but transcriptome changes
- transciprtome is linked to which genes which are expressed

### why rna
- medicine -> gene expression, tumors, disease linked to rna
- biology -> gene expression linked to cell type, embryonic development etc

### mRNA transcription

- exon is whats important
- introns is not transcribed 
- makes mRNA differ

## Rna-seq pipeline

- rna-seq is not actually sequencing rna, but cDNA from the RNA template
- reverse transcribed!
- reads are mapped to refrence genome
- *but...* we also wan to quantify, not just map

## FASTQ format

sequence id
base calls
seperator
quality score (q = quality, p = % base is wrong) (Q = -10log_10(p))

**example**

@SIM:1:FCX:1:15:6329:1045 1:N:0:2
ATGTAGTAGTG
+
<>;###


## What to do now

all samples have uniform dis. of reads = happy
*but...* 
samples will have different num. of reads

1. ## quality check the dataa
tidy up sequencing

use **FastQC** for this stuff
- quality control
- graphs/tables
- spits out html

use **trimmomatic** for adapter trimming
- adapter sequences are on the dna, and we want it gone
- remove other artifcats
- CLI to trim/crop illumina
- remove adapters
- works with paired/single ended reads (forward/reverse strand)
- fastq can be huge so compressed/decompressed

use **kraken** also its quality/trimming
- read many files at once
- uses PERL

### map to refrence genome
- rna seq is mapped to refrence
- FASTA files
- GTF files include genomic features, intron/exon
- example refrence genome is A FASTA file

### GTF
example:

1       havana  gene    11869   14409   .       +       .       gene_id "ENSG00000223972"; gene_version "5"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene";
1       havana  transcript      11869   14409   .       +       .       gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; transcript_name "DDX11L1-202"; transcript_source "havana"; transcript_biotype "processed_transcript"; tag "basic"; transcript_support_level "1";
1       havana  exon    11869   12227   .       +       .       gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; exon_number "1"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; transcript_name "DDX11L1-202"; transcript_source "havana"; transcript_biotype "processed_transcript"; exon_id "ENSE00002234944"; exon_version "1"; tag "basic"; transcript_support_level "1";

lots of info!
Info -> seq,source,feature,start,end,score,strand,frame,attribute

2. ## alignment 

Take all of FASTQ and fit them together

**STAR**

setup to align RNA reads to correct gene

get our reads, get our refrence genome
map this stuff!

*Mapping vs genome assembly*
with no refrence genome to map to, use genome assembly

### Star
- genome generation and mapping
- outputs sam/bam files and mapping reports
- sam is readble, bam is not


3. ## quanitfy alignment and downstream differences

**How much rna do we have? (abundance)**

fragments per kilobase per million
transcripts per million

**HTseq** is good at this using BAM/SAM and GTF
outputs a table of gene id's and counts

Ok cool now we know this but what now?

### Uses
- gene expression in sick people/vs control
- up/down regulation
- volcano plot can visualize gene clusters
- gene ontology enrichment (is a gene enriched for specific purpose?)